torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,227 @@
+    // and batching code, so leaving it out for now.
+    //
+    // auto curToken = g->addInput("world")->setType(WorldType::get());
+    // curToken = visitBlock(g->block(), curToken);
+    // g->registerOutput(curToken);
+
+    // Generate the first world token
+    const auto tokenGenerator = g->create(prim::Constant);
+    g->block()->prependNode(tokenGenerator);
+    auto curToken = tokenGenerator->output()->setType(WorldType::get());
+
+    visitBlock(g->block(), curToken);
+  }
+
+ private:
+  Value* visitBlock(Block* block, Value* curToken) {
+    for (auto* node : block->nodes()) {
+      curToken = visitNode(node, curToken);
+    }
+    return curToken;
+  }
+
+  // If a node uses a mutable variable (or mutates a previously constant
+  // variable), annotate it
+  //
+  // Returns the last world token emitted for subsequent annotations to use.
+  Value* visitNode(Node* node, Value* curToken) {
+    if (node->kind() == prim::MemoryFence) {
+      return curToken;
+    }
+
+    if (node->kind() == prim::If) {
+      JIT_ASSERT(node->blocks().size() == 2);
+
+      auto trueBlock = node->blocks().at(0);
+      auto falseBlock = node->blocks().at(1);
+
+      auto trueToken = visitBlock(trueBlock, curToken);
+      auto falseToken = visitBlock(falseBlock, curToken);
+
+      // If any branch has a mutating op, this node has to output a world token
+      if (trueToken != curToken || falseToken != curToken) {
+        trueBlock->registerOutput(trueToken);
+        falseBlock->registerOutput(falseToken);
+
+        return node->addOutput()->setType(WorldType::get());
+      }
+      return curToken;
+    }
+
+    if (node->kind() == prim::Loop) {
+      JIT_ASSERT(node->blocks().size() == 1);
+      auto block = node->blocks().at(0);
+      if (!shouldAnnotate(block)) {
+        // Bail out early if there's no mutable variables used inside
+        return curToken;
+      }
+
+      // Register the world token as a loop carried dependency
+      auto beginLoopToken = block->addInput()->setType(WorldType::get());
+      auto endLoopToken = visitBlock(block, beginLoopToken);
+      block->registerOutput(endLoopToken);
+
+      JIT_ASSERT(endLoopToken != beginLoopToken);
+
+      // Thread the world token through the loop node
+      node->addInput(curToken);
+      return node->addOutput()->setType(WorldType::get());
+    }
+
+    // For mutating ops, just thread the world token through the node.
+    if (isMutatingOp(node)) {
+      node->replaceInput(0, curToken);
+      return node->outputs().at(0);
+    }
+
+    JIT_ASSERT(node->blocks().size() == 0);
+
+    // For pure ops that need to be annotated, fence them.
+    if (shouldAnnotate(node)) {
+      return addFenceForNode(node, curToken);
+    }
+
+    return curToken;
+  }
+
+  bool shouldAnnotate(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (input->type()->kind() != TypeKind::ListType) {
```

Comment:
Maybe best to encapsulate this now as: `if(!isMutableType(input->type()))`. In the future I think we will likely keep the set of mutable types and immutable types separate.